File: ai agent app/agents/analyzer.py
"""agents/analyzer.py — the generator's compiler, as a Python client.

Wraps the long-lived `scripts/analyzer.mjs` process (a TypeScript LanguageService over stdio) so a
generation agent can ask "is this file correct?" BEFORE writing it. Measured on a real app: ~0.5-1.3s
per file, against 60-120s to generate one — the check is free, and it catches exactly the classes the
repair LLM could not fix (truncation, `</Alert}`, hallucinated icons, MUI v5 Grid).

Fails SOFT by design: if node or the app's typescript is missing, every call reports clean and
generation proceeds untouched. An analyzer that can't run must never be the reason an app doesn't get
built — the repair harness is still behind it.
"""
from __future__ import annotations

import json
import os
import queue
import shutil
import subprocess
import threading
from pathlib import Path

_NODE = shutil.which("node") or "node"
_SCRIPT = Path(__file__).resolve().parent.parent / "scripts" / "analyzer.mjs"

# The first check pays for building the program (~2.5s); later ones are ~1s. A file that takes longer
# than this is not worth stalling a 10-minute generation over.
_BOOT_TIMEOUT = 90.0
_CALL_TIMEOUT = 60.0


class Analyzer:
    """One analyzer process per generated app. Use as a context manager, or call close()."""

    def __init__(self, project_dir, emit=None):
        # Absolute: the child resolves argv[2] against its own cwd, which IS this directory — a
        # relative path resolves to `<dir>/<dir>`, tsconfig.json is not found, and the analyzer then
        # runs with an empty config (no `jsx`, no `paths`) and reports confident nonsense.
        self.dir = Path(project_dir).resolve()
        self.emit = emit or (lambda *a, **k: None)
        self.proc: subprocess.Popen | None = None
        self.ok = False
        self.ts_version = ""
        self._id = 0
        self._lines: queue.Queue = queue.Queue()
        self._lock = threading.Lock()
        self._start()

    # ── lifecycle ────────────────────────────────────────────────────────────
    def _start(self):
        if not _SCRIPT.exists():
            self._log(f"analyzer disabled: {_SCRIPT.name} not found")
            return
        try:
            self.proc = subprocess.Popen(
                [_NODE, str(_SCRIPT), str(self.dir)],
                stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL,
                text=True, encoding="utf-8", errors="replace", bufsize=1,
                cwd=str(self.dir),
                creationflags=getattr(subprocess, "CREATE_NO_WINDOW", 0) if os.name == "nt" else 0,
            )
        except Exception as e:  # noqa: BLE001
            self._log(f"analyzer disabled: could not start node ({e})")
            return

        threading.Thread(target=self._reader, daemon=True).start()
        hello = self._read(_BOOT_TIMEOUT)
        if not hello or not hello.get("ok"):
            reason = (hello or {}).get("error") or (hello or {}).get("fatal") or "no response"
            self._log(f"analyzer disabled: {reason}")
            self.close()
            return
        self.ok = True
        self.ts_version = hello.get("ts", "?")
        self._log(f"analyzer ready (typescript {self.ts_version}) — checking files before they land")

    def _reader(self):
        try:
            for line in self.proc.stdout:  # type: ignore[union-attr]
                line = line.strip()
                if line:
                    self._lines.put(line)
        except Exception:  # noqa: BLE001
            pass
        finally:
            self._lines.put(None)

    def _read(self, timeout: float):
        try:
            line = self._lines.get(timeout=timeout)
        except queue.Empty:
            return None
        if line is None:
            return None
        try:
            return json.loads(line)
        except Exception:  # noqa: BLE001
            return None

    def close(self):
        self.ok = False
        if self.proc:
            try:
                self.proc.stdin.close()  # type: ignore[union-attr]
                self.proc.wait(timeout=5)
            except Exception:  # noqa: BLE001
                try:
                    self.proc.kill()
                except Exception:  # noqa: BLE001
                    pass
            self.proc = None

    def __enter__(self):
        return self

    def __exit__(self, *_exc):
        self.close()

    def _log(self, text: str):
        self.emit("log", {"text": text})

    # ── protocol ─────────────────────────────────────────────────────────────
    def _call(self, op: str, **kw) -> dict | None:
        if not self.ok or not self.proc:
            return None
        with self._lock:
            self._id += 1
            req = {"id": self._id, "op": op, **kw}
            try:
                self.proc.stdin.write(json.dumps(req) + "\n")  # type: ignore[union-attr]
                self.proc.stdin.flush()  # type: ignore[union-attr]
            except Exception as e:  # noqa: BLE001
                self._log(f"analyzer lost ({e}) — continuing without it")
                self.ok = False
                return None
            res = self._read(_CALL_TIMEOUT)
            if res is None:
                self._log("analyzer timed out — continuing without it")
                self.ok = False
                return None
            return res

    # ── ops ──────────────────────────────────────────────────────────────────
    def check(self, rel: str, source: str) -> list[dict]:
        """Diagnostics for `source` as if it were at `rel` — WITHOUT writing it.

        Returns [] when clean or when the analyzer is unavailable (fail-soft: an absent analyzer must
        not block generation). Each diagnostic: {line, column, code, message}."""
        res = self._call("check", path=rel.replace("\\", "/"), source=source)
        if not res or not res.get("ok"):
            return []
        return res.get("diagnostics") or []

    def release(self, rel: str):
        """Drop the in-memory candidate for `rel` (it is now on disk, or abandoned)."""
        self._call("release", path=rel.replace("\\", "/"))

    def interface(self, rel: str) -> dict:
        """{name, props} of the file's default export, read from the AST — the fact a consumer needs
        so it stops guessing (`<Body recipe={x} />` against a body that takes no props)."""
        res = self._call("interface", path=rel.replace("\\", "/"))
        if not res or not res.get("ok"):
            return {}
        return {"name": res.get("name"), "props": res.get("props"), "found": res.get("found")}

    def exports(self, module: str) -> dict:
        """{exists, names} for a module specifier — `@mui/icons-material/Heart` does not exist,
        `Favorite` does, and the model cannot tell without being shown."""
        res = self._call("exports", module=module)
        if not res or not res.get("ok"):
            return {"exists": True, "names": []}  # unknown → do not fabricate a failure
        return {"exists": res.get("exists"), "names": res.get("names") or [],
                "reason": res.get("reason")}


def format_diagnostics(diags: list[dict], limit: int = 6, width: int = 320) -> str:
    """The compiler's own words, for the prompt that regenerates the file.

    Messages are capped: an overload-resolution failure can run ~900 chars of type dumps, which buys
    the model nothing and crowds out the file it is meant to be rewriting."""
    lines = []
    for d in diags[:limit]:
        loc = f"line {d['line']}" if d.get("line") else "?"
        msg = " ".join(str(d.get("message", "")).split())
        if len(msg) > width:
            msg = msg[:width].rstrip() + " …"
        lines.append(f"- {loc}: {msg} (TS{d.get('code')})")
    if len(diags) > limit:
        lines.append(f"- …and {len(diags) - limit} more")
    return "\n".join(lines)

File: ai agent app/scripts/analyzer.mjs
/**
 * scripts/analyzer.mjs — the generator's compiler.
 *
 * A long-lived process holding an incremental TypeScript LanguageService for ONE generated app, so
 * the generator can ask "is this file correct?" *before* writing it — instead of finding out from
 * `next build` ten minutes later, when only the repair loop can help and often can't.
 *
 * Cost, measured on a real app: ~2.5s to build the program once, then ~1s per file — against 60-120s
 * to generate that file. Checking is ~1% of writing.
 *
 * Protocol: JSON lines on stdin/stdout. One request per line, one response per line.
 *   → {"id":1,"op":"check","path":"components/pages/HomePage.tsx","source":"'use client'…"}
 *   ← {"id":1,"ok":true,"diagnostics":[{"line":12,"code":2769,"message":"No overload matches…"}]}
 *   → {"id":2,"op":"interface","path":"components/pages/HomePage.tsx"}
 *   ← {"id":2,"ok":true,"name":"HomePage","props":"{ initialItem: Record<string, any> }"}
 *   → {"id":3,"op":"exports","module":"@mui/icons-material/Favorite"}
 *   ← {"id":3,"ok":true,"exists":true,"names":["default"]}
 *
 * Usage: node scripts/analyzer.mjs <projectDir>
 * The app's OWN typescript is used (every generated app installs it), so diagnostics match the
 * `tsc --noEmit` the harness runs later, exactly.
 */
import fs from 'node:fs'
import path from 'node:path'
import readline from 'node:readline'
import { createRequire } from 'node:module'

const projectDir = path.resolve(process.argv[2] || '.')
const require = createRequire(path.join(projectDir, 'noop.js'))

let ts
try {
  ts = require('typescript') // the app's own compiler — same version the build will use
} catch {
  process.stdout.write(JSON.stringify({
    id: 0, ok: false, fatal: 'typescript-not-installed',
    error: `no typescript in ${projectDir}/node_modules — run the install before generating`,
  }) + '\n')
  process.exit(2)
}

// ── program state ────────────────────────────────────────────────────────────
// Overlays hold files the generator has NOT written yet: we check a candidate source in memory, so a
// bad file never reaches disk.
// Keys are POSIX-normalised: TypeScript hands the host forward-slash paths, so a Map keyed by
// Windows backslashes silently misses every lookup — the overlay is never found, the file reads as
// empty, and every check comes back "clean". Normalise on the way in and on every host callback.
const norm = (p) => path.resolve(projectDir, p).replace(/\\/g, '/')
const overlays = new Map() // normalised absolute path -> text
// Versions must be MONOTONIC and outlive their overlay: the document registry caches by
// (path, version), so recycling version 1 for a second candidate hands back the first one's AST —
// every file after the first gets checked as if it were the first.
const versions = new Map() // normalised absolute path -> int
const bump = (f) => {
  const v = (versions.get(f) || 0) + 1
  versions.set(f, v)
  return v
}
let rootNames = []
let options = {}

function loadConfig() {
  const configPath = path.join(projectDir, 'tsconfig.json')
  // No tsconfig means no `jsx` and no `paths`: every file would report "Cannot use JSX unless the
  // '--jsx' flag is provided" and "Cannot find module '@/types'". Refuse loudly — a misconfigured
  // analyzer that reports confident nonsense is worse than no analyzer at all.
  if (!fs.existsSync(configPath)) {
    process.stdout.write(JSON.stringify({
      id: 0, ok: false, fatal: 'no-tsconfig',
      error: `no tsconfig.json in ${projectDir} — cannot type-check without the app's own config`,
    }) + '\n')
    process.exit(2)
  }
  const read = ts.readConfigFile(configPath, ts.sys.readFile)
  const parsed = ts.parseJsonConfigFileContent(read.config || {}, ts.sys, projectDir)
  rootNames = parsed.fileNames.map(norm)
  options = parsed.options
}
loadConfig()

const host = {
  getScriptFileNames: () => [...new Set([...rootNames, ...overlays.keys()])],
  getScriptVersion: (f) => {
    const k = norm(f)
    if (versions.has(k)) return String(versions.get(k))
    try {
      return String(fs.statSync(f).mtimeMs)
    } catch {
      return '0'
    }
  },
  getScriptSnapshot: (f) => {
    const k = norm(f)
    if (overlays.has(k)) return ts.ScriptSnapshot.fromString(overlays.get(k))
    try {
      return ts.ScriptSnapshot.fromString(fs.readFileSync(f, 'utf8'))
    } catch {
      return undefined
    }
  },
  getCurrentDirectory: () => projectDir,
  getCompilationSettings: () => options,
  getDefaultLibFileName: (o) => ts.getDefaultLibFilePath(o),
  fileExists: (f) => overlays.has(norm(f)) || ts.sys.fileExists(f),
  readFile: (f) => (overlays.get(norm(f)) ?? ts.sys.readFile(f)),
  readDirectory: ts.sys.readDirectory,
  directoryExists: ts.sys.directoryExists,
  getDirectories: ts.sys.getDirectories,
  realpath: ts.sys.realpath,
}
const service = ts.createLanguageService(host, ts.createDocumentRegistry())

// ── ops ──────────────────────────────────────────────────────────────────────
const abs = norm

function fmt(d) {
  const out = { code: d.code, message: ts.flattenDiagnosticMessageText(d.messageText, ' ') }
  if (d.file && typeof d.start === 'number') {
    const { line, character } = d.file.getLineAndCharacterOfPosition(d.start)
    out.line = line + 1
    out.column = character + 1
  }
  return out
}

function check({ path: rel, source }) {
  const file = abs(rel)
  if (typeof source === 'string') {
    overlays.set(file, source)
    bump(file)
    if (!rootNames.includes(file)) rootNames.push(file)
  }
  // Syntactic first: a file that does not parse produces meaningless semantic noise.
  const syntactic = service.getSyntacticDiagnostics(file).map(fmt)
  if (syntactic.length) return { diagnostics: syntactic, kind: 'syntactic' }
  const semantic = service.getSemanticDiagnostics(file).map(fmt)
  return { diagnostics: semantic, kind: semantic.length ? 'semantic' : 'clean' }
}

/** Drop an overlay once the real file is on disk (or the candidate is abandoned). */
function release({ path: rel }) {
  const file = abs(rel)
  overlays.delete(file)
  bump(file) // the next read must come from disk, not the registry's cache of the overlay
  return { released: true }
}

/**
 * The default export's name and its exact props — what a consumer must be told so it stops guessing.
 * This is the fact behind the page/body prop mismatches: two agents agreeing by convention instead of
 * one reading the other.
 */
function iface({ path: rel }) {
  const file = abs(rel)
  // A file written after this process booted is not in the program yet — the composer asks for the
  // chunk's signature moments after the chunk was created, so without this it always answered "no
  // such file" and the page fell back to guessing the props.
  if (!rootNames.includes(file) && fs.existsSync(file)) {
    rootNames.push(file)
    bump(file)
  }
  const prog = service.getProgram()
  const sf = prog?.getSourceFile(file)
  if (!sf) return { name: null, props: null, found: false }
  let name = null
  let props = null
  const checker = prog.getTypeChecker()
  for (const st of sf.statements) {
    // `export default Name` is an ExportAssignment without a DefaultKeyword modifier.
    if (ts.isExportAssignment(st) && ts.isIdentifier(st.expression)) {
      name = st.expression.getText(sf)
      const declaration = sf.statements.find((candidate) =>
        ts.isFunctionDeclaration(candidate) && candidate.name?.text === name)
      const variable = sf.statements
        .filter((candidate) => ts.isVariableStatement(candidate))
        .flatMap((candidate) => [...candidate.declarationList.declarations])
        .find((candidate) => ts.isIdentifier(candidate.name) && candidate.name.text === name)
      const callable = declaration || ((variable?.initializer &&
        (ts.isArrowFunction(variable.initializer) || ts.isFunctionExpression(variable.initializer)))
        ? variable.initializer : null)
      const p = callable?.parameters?.[0]
      if (p) {
        const t = checker.getTypeAtLocation(p)
        props = p.type ? p.type.getText(sf) : checker.typeToString(t)
      } else if (callable?.parameters) {
        props = ''
      }
      continue
    }
    const isDefault = st.modifiers?.some((m) => m.kind === ts.SyntaxKind.DefaultKeyword)
    if (!isDefault) continue
    if (ts.isFunctionDeclaration(st)) {
      name = st.name?.getText(sf) ?? null
      const p = st.parameters?.[0]
      if (p) {
        const t = checker.getTypeAtLocation(p)
        props = p.type ? p.type.getText(sf) : checker.typeToString(t)
      } else {
        props = '' // takes no props — say so explicitly, it is the fact consumers need
      }
    } else if (ts.isExportAssignment(st) && ts.isIdentifier(st.expression)) {
      name = st.expression.getText(sf)
    }
  }
  return { name, props, found: true }
}

/**
 * What a module REALLY exports. `@mui/icons-material/Heart` does not exist; `Favorite` does — and the
 * model cannot tell without being shown. A file check is enough for icons; anything else is parsed.
 */
function exportsOf({ module: spec }) {
  // Bare package (possibly with a subpath) → resolve inside node_modules.
  if (!spec.startsWith('.') && !spec.startsWith('@/')) {
    const parts = spec.split('/')
    const pkg = spec.startsWith('@') ? parts.slice(0, 2).join('/') : parts[0]
    const sub = spec.slice(pkg.length + 1)
    const base = path.join(projectDir, 'node_modules', ...pkg.split('/'))
    if (!fs.existsSync(base)) return { exists: false, reason: 'package-not-installed', names: [] }
    if (!sub) return { exists: true, names: [] }
    const candidates = [`${sub}.js`, `${sub}.d.ts`, path.join(sub, 'index.js'), sub]
    const hit = candidates.some((c) => fs.existsSync(path.join(base, c)))
    return { exists: hit, reason: hit ? null : 'subpath-not-found', names: hit ? ['default'] : [] }
  }
  // Internal module → ask the compiler what it exports.
  const rel = spec.startsWith('@/') ? spec.slice(2) : spec
  const prog = service.getProgram()
  for (const ext of ['.ts', '.tsx', '/index.ts', '/index.tsx', '']) {
    const sf = prog?.getSourceFile(abs(rel + ext))
    if (!sf) continue
    const sym = prog.getTypeChecker().getSymbolAtLocation(sf)
    const names = sym ? prog.getTypeChecker().getExportsOfModule(sym).map((s) => s.getName()) : []
    return { exists: true, names }
  }
  return { exists: false, reason: 'module-not-found', names: [] }
}

const OPS = { check, interface: iface, exports: exportsOf, release, ping: () => ({ pong: true }) }

// ── stdio loop ───────────────────────────────────────────────────────────────
const rl = readline.createInterface({ input: process.stdin })
rl.on('line', (line) => {
  if (!line.trim()) return
  let req
  try {
    req = JSON.parse(line)
  } catch (e) {
    process.stdout.write(JSON.stringify({ id: null, ok: false, error: `bad json: ${e.message}` }) + '\n')
    return
  }
  const op = OPS[req.op]
  if (!op) {
    process.stdout.write(JSON.stringify({ id: req.id, ok: false, error: `unknown op: ${req.op}` }) + '\n')
    return
  }
  try {
    process.stdout.write(JSON.stringify({ id: req.id, ok: true, ...op(req) }) + '\n')
  } catch (e) {
    process.stdout.write(JSON.stringify({ id: req.id, ok: false, error: String(e?.message || e) }) + '\n')
  }
})
rl.on('close', () => process.exit(0))
process.stdout.write(JSON.stringify({ id: 0, ok: true, ready: true, ts: ts.version }) + '\n')
